README.md
[![Build Status](https://travis-ci.org/postmanlabs/postman-collection-transformer.svg?branch=develop)](https://travis-ci.org/postmanlabs/postman-collection-transformer) [![Greenkeeper badge](https://badges.greenkeeper.io/postmanlabs/postman-collection-transformer.svg)](https://greenkeeper.io/)
# postman-collection-transformer
Perform rapid conversion of JSON structure between Postman Collection Format v1 and v2.

The formats are documented at https://schema.getpostman.com

## Installation

For CLI usage:

    $ npm install -g postman-collection-transformer

As a library:

    $ npm install --save postman-collection-transformer

## Usage

#### Converting Entire Collections

The transformer provides a Command line API to convert collections.

Example:

    $ transform-collection convert \
        --input ./v1-collection.json \
        --input-version 2.0.0 \
        --output ./v2-collection.json \
        --output-version 1.0.0 \
        --pretty \
        --overwrite

All options:

    $ transform-collection convert -h

      Usage: convert [options]

      Convert Postman Collection from one format to another

      Options:

        -h, --help                      output usage information
        -i, --input <path>              path to the input postman collection file
        -j, --input-version [version]   the version of the input collection format standard (v1 or v2)
        -o, --output <path>             target file path where the converted collection will be written
        -p, --output-version [version]  required version to which the collection is needed to be converted to
        -P, --pretty                    Pretty print the output
        --retain-ids                    Retain the request and folder IDs during conversion (collection ID is always retained)
        -w, --overwrite                 Overwrite the output file if it exists

If you'd rather use the transformer as a library:
```javascript
    var transformer = require('postman-collection-transformer'),
        collection = require('./path/to/collection.json'),
        inspect = require('util').inspect,

        options = {
            inputVersion: '1.0.0',
            outputVersion: '2.0.0',
            retainIds: true  // the transformer strips request-ids etc by default.
        };

    transformer.convert(collection, options, function (error, result) {
        if (error) {
            return console.error(error);
        }

        // result <== the converted collection as a raw Javascript object
        console.log(inspect(result, {colors: true, depth: 10000}));
    });
```
    
#### Converting Individual Requests

The transformer also allows you to convert individual requests (only supported when used as a library):

###### Example
```javascript

    var transformer = require('postman-collection-transformer'),
    
        objectToConvert = { /* A valid collection v1 Request or a collection v2 Item */ },

        options = {
            inputVersion: '1.0.0',
            outputVersion: '2.0.0',
            retainIds: true  // the transformer strips request-ids etc by default.
        };

    transformer.convertSingle(objectToConvert, options, function (err, converted) {
        console.log(converted);
    });
```

#### Converting Individual Responses

You can convert individual responses too if needed:

###### Example
```javascript

    var transformer = require('postman-collection-transformer'),
    
        objectToConvert = { /* A v1 Response or a v2 Response */ },

        options = {
            inputVersion: '1.0.0',
            outputVersion: '2.0.0',
            retainIds: true  // the transformer strips request-ids etc by default.
        };

    transformer.convertResponse(objectToConvert, options, function (err, converted) {
        console.log(converted);
    });
```

#### Normalizing v1 collections

The transformer also provides a Command line API to normalize collections for full forward compatibility.

Example:

    $ transform-collection normalize \
        --input ./v1-collection.json \
        --normalize-version 1.0.0 \
        --output ./v1-norm-collection.json \
        --pretty \
        --overwrite

All options:

    $ transform-collection normalize -h

      Usage: normalize [options]

      Normalizes a postman collection according to the provided version


      Options:

        -i, --input <path>                 Path to the collection JSON file to be normalized
        -n, --normalize-version <version>  The version to normalizers the provided collection on
        -o, --output <path>                Path to the target file, where the normalized collection will be written
        -P, --pretty                       Pretty print the output
        --retain-ids                       Retain the request and folder IDs during conversion (collection ID is always retained)
        -w, --overwrite                    Overwrite the output file if it exists
        -h, --help                         Output usage information


If you'd rather use the transformer as a library:
```javascript
    var transformer = require('postman-collection-transformer'),
        collection = require('./path/to/collection.json'),
        inspect = require('util').inspect,

        options = {
            normalizeVersion: '1.0.0',
            mutate: false, // performs in-place normalization, false by default.
            noDefaults: false, // when set to true, sensible defaults for missing properties are skipped. Default: false
            prioritizeV2: false, // when set to true, v2 attributes are used as the source of truth for normalization.
            retainEmptyValues: false, // when set to true, empty values are set to '', not removed. False by default.
            retainIds: true  // the transformer strips request-ids etc by default.
        };

    transformer.normalize(collection, options, function (error, result) {
        if (error) {
            return console.error(error);
        }

        // result <== the converted collection as a raw Javascript object
        console.log(inspect(result, {colors: true, depth: 10000}));
    });
```


CHANGELOG.md
# Postman Collection Transformer Changelog

#### v3.0.0 (November 8, 2018)
* Updated dependencies :arrow_up:
* #112 Corrected retainIds behaviour for v1->v2 transformations
* #121 Added handling for null valued request bodies
* #126 Switched to Travis for Windows tests
* #124 Dropped support for Node v4.
* #123 Removed NSP (replaced with Snyk)
* #122 Added `.gitattributes` to enforce consistent line endings

#### v2.8.0 (September 20, 2018)
* #118 Added support for the `protocolProfileBehaviour` data element

#### v2.7.0 (September 11, 2018)
* #117 Added support for disabled request body flag

#### v2.6.3 (August 29, 2018)
* #115 Coerced non-string formdata file values in request bodies to `null`

#### v2.6.2 (August 6, 2018)
* #111 Updated `url.parse` to correctly account for path variables

#### v2.6.1 (July 24, 2018)
* #110 Prevent folder id replacement in v2.x -> v1 transformations :bug:
* #109 Increased test coveragee to 95% :arrow_up:

#### v2.6.0 (July 10, 2018)
* #107 Added support for `retainEmptyValues` in **all** flows :tada:.

#### v2.5.10 (July 4, 2018)
* #106 Updated dependencies :arrow_up:
* #105 Fixed a bug in url parsing protocol extraction :bug:

#### v2.5.9 (April 30, 2018)
* #103 Fixed `retainIds` behaviour for v1 to v2 conversions :bug:
* #104 Updated dependencies :arrow_up:

#### v2.5.8 (March 13, 2018)
* :arrow_up: Updated dependencies
* #102 Improved handling for `null` valued query parameters in transformations from `v1`

#### v2.5.7 (March 1, 2018)
* #101 Fixed text `type` mapping for `auth` and `variable` fields. :bug:

#### v2.5.6 (February 23, 2018)
* #100 Auth mapping improvements

#### v2.5.5 (February 6, 2018)
* :arrow_up: Updated dependencies
* :bug: Fixed empty description bug for v1 -> v2 transformations #99

#### v2.5.4 (January 7, 2018)
* :arrow_up: Updated dependencies
* #98 Improved `script.exec` type safety in v1 normalization :bug:

#### v2.5.3 (December 18, 2017)
* #97 Fixed a bug that caused query params to be duplicted :bug:

#### v2.5.2 (December 12, 2017)
* #96 Made empty description pruning optional with `retainEmptyValues` :bug:

#### v2.5.1 (December 1, 2017)
* #95 Improved type safety in folder and request transformations :bug:

#### v2.5.0 (November 22, 2017)
* #94 Added support for the `prioritizeV2` option to v1 normalization and v1 -> v2.x conversions. :tada:

#### v2.4.3 (November 20, 2017)
* #92 Fixed legacy property handling for non-complete event normalization :bug:

#### v2.4.2 (November 20, 2017)
* #91 Fixed default auth state derivation logic :bug:

#### v2.4.1 (November 18, 2017)
* :arrow_up: Updated dependencies #88
* :bug: Fixed empty stage handling logic for auth helpers #90
* :art: Reorganized package structure #89

#### v2.4.0 (November 7, 2017)
* #87 Tested code on Node v8 :tada:
* #84 Added support for `noDefaults` for v1 normalizations :tada:
* #82 `noauth` and `normal` auth handlers are now left alone :bug:
* #81 Added support for `mutate` to v1 normalizations :tada:
* #79 Added `normalize.*` function to sanitize v1 collections. :tada:
* #78 Added support for inherited entity transformations :tada:

#### v2.3.1 (October 3, 2017)
* :bug: Corrected OAuth2 param transformations #77

#### v2.3.0 (September 30, 2017)
* :tada: Added support for NTLM and Bearer Token auth transformations #76

#### v2.2.1 (September 28, 2017)
* Restored support for string-object hybrid v2 URLs #75

#### v2.2.0 (September 4, 2017)
* :tada: Added support for transformations to and from the `v2.1.0` format. #66, #67, #68
* :arrow_up: Updated dependencies #65
* :bug: Enforced an object structure for V2 collection request URLs. #62
* :bug: Prevented empty descriptions from showing up across transformations. #61

#### 2.1.5 (June 28, 2017)
* Allowed both: `key` and `id` as identifers for path variables in v2 collections #60

#### 2.1.4 (June 19, 2017)
* No code changes

#### 2.1.3 (June 09, 2017)
* Fixed a bug in the URL parse implementation, by copying the parser & unparser from the sdk

#### 2.1.2 (May 25, 2017)
* Handle an edge case where an unhandled ID is provided in the `requestObject`

#### 2.1.1 (May 22, 2017)
* Fixed a bug where the `requestObject` was missing when doing v2 to v1 conversions
* Added the ability to convert single responses as well

#### 2.1.0 (May 12, 2017)
* Descriptions for headers, url query params, url path variables, and request body data are now handled in transformations.
* Fix disabled: true header (`v2`) <-> // (commented) header (`v1`) transformations.
* Fixed bug where both enabled and disabled would appear in transformed entities.
* Request body transformations now follow `v1` - `v2` compliance
* Removed unwanted `enabled: true` properties from query params, headers, and request body data


lib/util.js
/* eslint-disable object-shorthand */
var _ = require('lodash'),
    url = require('./url'),
    rnd = Math.random;

module.exports = {
    // @todo: Add support for a `json` type once it becomes available
    typeMap: {
        string: 'string',
        boolean: 'boolean',
        number: 'number'
    },

    /**
     * Returns unique GUID on every call as per pseudo-number RFC4122 standards.
     *
     * @type {function}
     * @returns {string}
     */
    uid: function () {
        var n,
            r,
            E = '',
            H = '-'; // r = result , n = numeric variable for positional checks

        // if "n" is not 9 or 14 or 19 or 24 return a random number or 4
        // if "n" is not 15 generate a random number from 0 to 15
        // `(n ^ 20 ? 16 : 4)` := unless "n" is 20, in which case a random number from 8 to 11 otherwise 4
        //
        // in other cases (if "n" is 9,14,19,24) insert "-"
        // eslint-disable-next-line curly
        for (r = n = E; n++ < 36; r += n * 51 & 52 ? (n ^ 15 ? 8 ^ rnd() * (n ^ 20 ? 16 : 4) : 4).toString(16) : H);

        return r;
    },

    urlparse: function (u) {
        return url.parse(u);
    },

    urlunparse: function (urlObj) {
        return url.unparse(urlObj);
    },

    /**
     * A generic utility method to sanitize variable transformations across collection formats.
     *
     * @param {Object} entity - A generic object that could contain variable data.
     * @param {?Object} options - The set of options for variable handling.
     * @param {?Object} options.fallback - The fallback values to be used if no variables are present.
     * @param {?Boolean} options.noDefaults - When set to true, id will be retained.
     * @param {?Boolean} options.retainEmptyValues - When set to true, empty property values will be set to ''
     * instead of being deleted.
     * @param {Object} [modifiers] - A set of behavioral modifiers for variable handling.
     * @param {Boolean} [modifiers.isV1=false] - When set to true, looks for the pathVariableData property as well.
     * @returns {Object[]} - The set of sanitized entity level variables.
     */
    handleVars: function (entity, options, modifiers) {
        !options && (options = {});

        var self = this,
            noDefaults = options.noDefaults,
            isV1 = modifiers && modifiers.isV1,
            retainEmpty = options.retainEmptyValues,
            source = entity && (entity.variables || entity.variable || (isV1 && entity.pathVariableData)),
            fallback = options.fallback && options.fallback.values,
            result = _.map(source || fallback, function (item) {
                var result = {
                    id: (noDefaults || item.id) ? item.id : self.uid(),
                    key: item.key || item.id,
                    value: item.value,
                    type: (item.type === 'text' ? 'string' : item.type) || self.typeMap[typeof item.value] || 'any'
                };

                item.disabled && (result.disabled = true);

                if (item.description) { result.description = item.description; }
                else if (retainEmpty) { result.description = null; }

                return result;
            });

        if (result.length) { return result; }
    },

    /**
     * Performs auth cleansing common to all sorts of auth transformations.
     *
     * @param {Object} entity - The wrapped auth entity to be cleaned.
     * @param {?Object} options - The set of options for the current auth cleansing operation.
     * @param {?Boolean} [options.excludeNoauth=false] - When set to true, noauth is set to null.
     * @returns {Object|*} - The processed auth data.
     */
    cleanAuth: function (entity, options) {
        !options && (options = {});

        var auth = entity && entity.auth;

        if (auth === null) { return null; } // eslint-disable-line security/detect-possible-timing-attacks
        if (!(auth && auth.type)) { return; }
        if (auth.type === 'noauth') {
            return options.excludeNoauth ? null : { type: 'noauth' };
        }

        return auth;
    },

    cleanEmptyValue: function (entity, property, retainEmpty) {
        if (_.has(entity, property) && !entity[property]) {
            retainEmpty ? (entity[property] = null) : (delete entity[property]);
        }

        return entity;
    },

    /**
     * Transforms an array of auth params to their object equivalent.
     *
     * @param {Object} entity - The wrapper object for the array of auth params.
     * @param {?Object} options - The set of options for the current auth cleansing operation.
     * @param {?Boolean} [options.excludeNoauth=false] - When set to true, noauth is set to null.
     * @returns {*}
     */
    authArrayToMap: function (entity, options) {
        var type,
            result,
            self = this,
            auth = self.cleanAuth(entity, options);

        if (!auth) { return auth; }

        result = { type: (type = auth.type) };

        if (type !== 'noauth') {
            result[type] = _.transform(auth[type], function (result, param) {
                result[param.key] = param.value;
            }, {});
        }

        return result;
    },

    /**
     * Transforms an object of auth params to their array equivalent.
     *
     * @param {Object} entity - The wrapper object for the array of auth params.
     * @param {?Object} options - The set of options for the current auth cleansing operation.
     * @param {?Boolean} [options.excludeNoauth=false] - When set to true, noauth is set to null.
     * @returns {*}
     */
    authMapToArray: function (entity, options) {
        var type,
            params,
            result,
            self = this,
            auth = self.cleanAuth(entity, options);

        if (!auth) { return auth; }

        result = { type: (type = auth.type) };

        if (type !== 'noauth') {
            // @todo: Handle all non _ prefixed properties, ala request bodies
            params = _.map(auth[type], function (value, key) {
                return {
                    key: key,
                    value: value,
                    type: self.typeMap[typeof value] || 'any'
                };
            });

            params.length && (result[type] = params);
        }

        return result;
    },

    /**
     * Sanitizes a collection SDK compliant auth list.
     *
     * @param {Object} entity - The wrapper entity for the auth manifest.
     * @param {?Object} options - The set of options for the current auth cleansing operation.
     * @param {?Boolean} [options.excludeNoauth=false] - When set to true, noauth is set to null.
     * @returns {Object[]} - An array of raw collection SDK compliant auth parameters.
     */
    sanitizeAuthArray: function (entity, options) {
        var type,
            result,
            self = this,
            auth = self.cleanAuth(entity, options);

        if (!auth) { return auth; }

        result = { type: (type = auth.type) };

        if (type !== 'noauth') {
            result[type] = _.map(auth[type], function (param) {
                return {
                    key: param.key,
                    value: param.value,
                    type: (param.type === 'text' ? 'string' : param.type) || self.typeMap[typeof param.value] || 'any'
                };
            });
        }

        return result;
    },

    /**
     * A helper function to determine if the provided v1 entity has legacy properties.
     *
     * @private
     * @param {Object} entityV1 - The v1 entity to be checked for the presence of legacy properties.
     * @param {String} type - The type of property to be adjudged against.
     * @returns {Boolean|*} - A flag to indicate the legacy property status of the passed v1 entity.
     */
    notLegacy: function (entityV1, type) {
        if (!entityV1) { return; }

        switch (type) {
            case 'event':
                return !(entityV1.tests || entityV1.preRequestScript);
            case 'auth':
                return _.has(entityV1, 'auth') && !(_.has(entityV1, 'currentHelper') || entityV1.helperAttributes);
            default:
                return true;
        }
    },

    authMappersFromLegacy: {
        basicAuth: function (oldParams) {
            return oldParams && {
                username: oldParams.username,
                password: oldParams.password,
                saveHelperData: oldParams.saveToRequest,
                showPassword: false
            };
        },
        bearerAuth: function (oldParams) {
            return oldParams && {
                token: oldParams.token
            };
        },
        digestAuth: function (oldParams) {
            return oldParams && {
                algorithm: oldParams.algorithm,
                username: oldParams.username,
                realm: oldParams.realm,
                password: oldParams.password,
                nonce: oldParams.nonce,
                nonceCount: oldParams.nonceCount,
                clientNonce: oldParams.clientNonce,
                opaque: oldParams.opaque,
                qop: oldParams.qop,
                disableRetryRequest: oldParams.disableRetryRequest
            };
        },
        oAuth1: function (oldParams) {
            return oldParams && {
                consumerKey: oldParams.consumerKey,
                consumerSecret: oldParams.consumerSecret,
                token: oldParams.token,
                tokenSecret: oldParams.tokenSecret,
                signatureMethod: oldParams.signatureMethod,
                timestamp: oldParams.timestamp,
                nonce: oldParams.nonce,
                version: oldParams.version,
                realm: oldParams.realm,
                addParamsToHeader: oldParams.header,
                autoAddParam: oldParams.auto,
                addEmptyParamsToSign: oldParams.includeEmpty
            };
        },
        hawkAuth: function (oldParams) {
            return oldParams && {
                authId: oldParams.hawk_id,
                authKey: oldParams.hawk_key,
                algorithm: oldParams.algorithm,
                user: oldParams.user,
                saveHelperData: oldParams.saveToRequest,
                nonce: oldParams.nonce,
                extraData: oldParams.ext,
                appId: oldParams.app,
                delegation: oldParams.dlg,
                timestamp: oldParams.timestamp
            };
        },
        ntlmAuth: function (oldParams) {
            return oldParams && {
                username: oldParams.username,
                password: oldParams.password,
                domain: oldParams.domain,
                workstation: oldParams.workstation,
                disableRetryRequest: oldParams.disableRetryRequest
            };
        },
        oAuth2: function (oldParams) {
            return oldParams && {
                accessToken: oldParams.accessToken,
                addTokenTo: oldParams.addTokenTo,
                callBackUrl: oldParams.callBackUrl,
                authUrl: oldParams.authUrl,
                accessTokenUrl: oldParams.accessTokenUrl,
                clientId: oldParams.clientId,
                clientSecret: oldParams.clientSecret,
                clientAuth: oldParams.clientAuth,
                grantType: oldParams.grantType,
                scope: oldParams.scope,
                username: oldParams.username,
                password: oldParams.password,
                tokenType: oldParams.tokenType,
                redirectUri: oldParams.redirectUri,
                refreshToken: oldParams.refreshToken
            };
        },
        // Only exists for consistency
        awsSigV4: function (oldParams) {
            return oldParams;
        }
    },
    authMappersFromCurrent: {
        basicAuth: function (newParams) {
            return newParams && {
                id: 'basic',
                username: newParams.username,
                password: newParams.password,
                saveToRequest: newParams.saveHelperData
            };
        },
        bearerAuth: function (newParams) {
            return newParams && {
                id: 'bearer',
                token: newParams.token
            };
        },
        digestAuth: function (newParams) {
            return newParams && {
                id: 'digest',
                algorithm: newParams.algorithm,
                username: newParams.username,
                realm: newParams.realm,
                password: newParams.password,
                nonce: newParams.nonce,
                nonceCount: newParams.nonceCount,
                clientNonce: newParams.clientNonce,
                opaque: newParams.opaque,
                qop: newParams.qop,
                disableRetryRequest: newParams.disableRetryRequest
            };
        },
        oAuth1: function (newParams) {
            return newParams && {
                id: 'oAuth1',
                consumerKey: newParams.consumerKey,
                consumerSecret: newParams.consumerSecret,
                token: newParams.token,
                tokenSecret: newParams.tokenSecret,
                signatureMethod: newParams.signatureMethod,
                timestamp: newParams.timeStamp || newParams.timestamp,
                nonce: newParams.nonce,
                version: newParams.version,
                realm: newParams.realm,
                header: newParams.addParamsToHeader,
                auto: newParams.autoAddParam,
                includeEmpty: newParams.addEmptyParamsToSign
            };
        },
        hawkAuth: function (newParams) {
            return newParams && {
                id: 'hawk',
                hawk_id: newParams.authId,
                hawk_key: newParams.authKey,
                algorithm: newParams.algorithm,
                user: newParams.user,
                saveToRequest: newParams.saveHelperData,
                nonce: newParams.nonce,
                ext: newParams.extraData,
                app: newParams.appId,
                dlg: newParams.delegation,
                timestamp: newParams.timestamp
            };
        },
        ntlmAuth: function (newParams) {
            return newParams && {
                id: 'ntlm',
                username: newParams.username,
                password: newParams.password,
                domain: newParams.domain,
                workstation: newParams.workstation,
                disableRetryRequest: newParams.disableRetryRequest
            };
        },
        oAuth2: function (newParams) {
            return newParams && {
                id: 'oAuth2',
                accessToken: newParams.accessToken,
                addTokenTo: newParams.addTokenTo,
                callBackUrl: newParams.callBackUrl,
                authUrl: newParams.authUrl,
                accessTokenUrl: newParams.accessTokenUrl,
                clientId: newParams.clientId,
                clientSecret: newParams.clientSecret,
                clientAuth: newParams.clientAuth,
                grantType: newParams.grantType,
                scope: newParams.scope,
                username: newParams.username,
                password: newParams.password,
                tokenType: newParams.tokenType,
                redirectUri: newParams.redirectUri,
                refreshToken: newParams.refreshToken
            };
        },
        // Only exists for consistency
        awsSigV4: function (newParams) {
            return newParams;
        }
    },

    /**
     * Validate protocolProfileBehavior property's value.
     *
     * @param {Object} source - A generic object that could contain the protocolProfileBehavior property.
     * @param {?Object} destination - The destination object that needs the addition of protocolProfileBehavior.
     * @returns {Boolean} - A Boolean value to decide whether to include the property or not.
     */
    addProtocolProfileBehavior: function (source, destination) {
        var behavior = source && source.protocolProfileBehavior;

        if (!(behavior && typeof behavior === 'object')) { return false; }

        destination && (destination.protocolProfileBehavior = behavior);

        return true;
    }
};
